java/cuvs-java/src/main/java22/com/nvidia/cuvs/internal/TieredIndexImpl.java

```diff
@@ -134,35 +135,33 @@ private IndexReference build() {
 
         // TieredIndex REQUIRES device memory - allocate it
         long datasetSize = C_FLOAT_BYTE_SIZE * rows * cols;
-        MemorySegment datasetDP = allocateRMMSegment(cuvsRes, datasetSize);
+        try (var datasetDP = allocateRMMSegment(cuvsRes, datasetSize)) {
+          // Copy host to device
+          Util.cudaMemcpy(datasetDP.handle(), hostDataSeg, datasetSize, HOST_TO_DEVICE);
 
-        // Copy host to device
-        Util.cudaMemcpy(datasetDP, hostDataSeg, datasetSize, HOST_TO_DEVICE);
+          // Create tensor from device memory
+          long[] datasetShape = {rows, cols};
+          MemorySegment datasetTensor =
+              prepareTensor(
+                  localArena, datasetDP.handle(), datasetShape, kDLFloat(), 32, kDLCUDA(), 1);
 
-        // Create tensor from device memory
-        long[] datasetShape = {rows, cols};
-        MemorySegment datasetTensor =
-            prepareTensor(localArena, datasetDP, datasetShape, kDLFloat(), 32, kDLCUDA(), 1);
+          MemorySegment index = localArena.allocate(cuvsTieredIndex_t);
```

Comment:
Unrelated to this PR, but I'm conflicted about using cuvsTieredIndex_t, since we need a POINTER here (as we see below, when we are extracting a C_POINTER).
Right now, cuvsTieredIndex_t is a pointer, but if that changes, we would end up allocating e.g. a struct, and use it to store (and retrieve) a pointer...

Reply:
> Right now, `cuvsTieredIndex_t` is a pointer, but if that changes...

I think that could apply to any of the typedefs exposed in the C++ API.  Changing those structures would be a non-trivial event.  I would think @benfred and gang are likely to keep those stable. :]